examples/chain_replication.pyv

```diff
@@ -0,0 +1,89 @@
+# node ids and states
+sort node_id
+sort node_state
+
+# ordering on node states
+immutable relation le(node_state, node_state)
+axiom le(X,X) # reflexive
+axiom le(X,Y) & le(Y,X) -> X = Y # anti-symmetric
+axiom le(X,Y) & le(Y,Z) -> le(X,Z) # transitive
+axiom le(X,Y) | le(Y,X) # total
+
+# empty state
+immutable constant empty: node_state
+
+# the empty state is the least state
+axiom le(empty, X)
+
+# let's say that there are three nodes
+immutable constant head: node_id
+immutable constant middle: node_id
+immutable constant tail: node_id
+
+# the head, middle, and tail nodes are all different
+# (this is too strong; it should be OK for them to all be the same, too)
+axiom head != middle & head != tail & middle != tail
+
+# nodes have states
+# TODO: maybe this could just be a big mutable relation
+mutable constant head_state : node_state
+mutable constant middle_state : node_state
+mutable constant tail_state : node_state
+
+# set of messages in transit: sender, recipient, data
+mutable relation msg(node_id, node_id, node_state)
+
+# all replicas start in the empty state
+init head_state = empty
+init middle_state = empty
+init tail_state = empty
+# with no messages in transit
+init !msg(X, Y, Z)
+
+# the head replica can move to a new state and send its state to the middle node
+transition head_send(next: node_state)
+  modifies head_state, msg
+  le(head_state, next) & # the new state must be bigger
+  new(head_state) = next &
+  # add next to the set of messages sent from head to middle
+  (new(msg(X, Y, Z)) <-> msg(X, Y, Z) | (X = head & Y = middle & Z = next))
+
+# the middle replica can receive a message and update its state
+transition middle_recv(recvd: node_state)
+  modifies middle_state, msg
+  le(middle_state, recvd) &
+  msg(head, middle, recvd) & # the received message must have come from head
+  (msg(head, middle, Z) -> le(recvd, Z)) & # this must be the smallest of the messages from head
+  new(middle_state) = recvd &
+  # take recvd out of the set of messages
+  (new(msg(X, Y, Z)) <-> msg(X, Y, Z) & (X = head & Y = middle & Z != recvd))
```

Comment:
I think this doesn't quite do what you want. I believe the line should read

```
  (new(msg(X, Y, Z)) <-> msg(X, Y, Z) & !(X = head & Y = middle & Z = recvd))
```

which is the idiomatic way to remove the tuple (head, middle, recvd) from the relation msg. What's there now does something more, namely removing *all* tuples in the relation except those from head to middle that are not the one received.

To demonstrate this, consider the following trace query

```
sat trace {
  head_send
  middle_recv
  middle_send
  assert exists S. msg(middle, tail, S)
}
```

which asks whether it is possible that there are messages in the network from middle to tail after `middle_send`. mypyvy reports sat, which makes sense. However, if we then let the head node update its state again, and forward to middle, and middle receives it, the line highlighted above will incorrectly remove the message from middle to tail!

```
sat trace {
  head_send
  middle_recv
  middle_send
  assert exists S. msg(middle, tail, S)
  head_send
  middle_recv
  assert exists S. msg(middle, tail, S)
}
```

Intuitively, this query should be sat, but it is unsat in the current model.

Reply:
Oops, thanks for catching that!